pkg/guestagent/events/containerd_linux.go

```diff
@@ -0,0 +1,369 @@
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/lima-vm/lima/v2/pkg/guestagent/api"
+)
+
+const (
+	stateKey             = "nerdctl/state-dir"
+	portsKey             = "nerdctl/ports"
+	namespaceKey         = "nerdctl/namespace"
+	defaultSocketTimeout = 5 * time.Second
+)
+
+type ContainerdEventMonitor struct {
+	socketPaths            []string
+	runningContainersMutex sync.Mutex
+	runningContainers      map[string][]*api.IPPort
+}
+
+func NewContainerdEventMonitor(socketPaths []string) *ContainerdEventMonitor {
+	return &ContainerdEventMonitor{
+		socketPaths:       socketPaths,
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (c *ContainerdEventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event, errCh chan error) {
+	var wg sync.WaitGroup
+	for _, socket := range c.socketPaths {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					errCh <- ctx.Err()
+					return
+				default:
+				}
+				info, err := os.Stat(socket)
+				if err != nil {
+					if os.IsNotExist(err) {
+						logrus.Warnf("containerd socket %s does not exist", socket)
+						// Wait for 2s before retrying again
+						time.Sleep(2 * time.Second)
+					} else {
+						logrus.Errorf("failed to stat containerd socket %s: %v", socket, err)
+					}
+					continue
+				}
+				if info.IsDir() {
+					errCh <- fmt.Errorf("containerd socket path %s is a directory", socket)
+					// this is unrecoverable
+					return
+				}
+				cli, err := containerd.New(socket, containerd.WithDefaultNamespace(containerdNamespace.Default))
+				if err != nil {
+					logrus.Warnf("failed to create client for socket %s: %v", socket, err)
+					continue
+				}
+				clientCtx, cancel := context.WithTimeout(ctx, defaultSocketTimeout)
+				serving, serveErr := cli.IsServing(clientCtx)
+				cancel()
+				if serveErr != nil || !serving {
+					logrus.Warnf("containerd daemon not serving on socket %s: %v. Retrying in 5s...", socket, serveErr)
+					cli.Close()
+					time.Sleep(5 * time.Second)
+					continue
+				}
+				logrus.Infof("successfully connected to containerd on socket %s", socket)
+				if err := c.runMonitorClient(ctx, cli, ch); err != nil {
+					logrus.Errorf("containerd port monitoring for socket: %s failed: %s", socket, err)
+					errCh <- err
+				}
+				cli.Close()
+			}
+		}(socket)
+	}
+	wg.Wait()
+}
+
+func (c *ContainerdEventMonitor) runMonitorClient(ctx context.Context, cli *containerd.Client, ch chan *api.Event) error {
+	subscribeFilters := []string{
+		`topic=="/tasks/start"`,
+		`topic=="/containers/update"`,
+		`topic=="/tasks/exit"`,
+	}
+	msgCh, errCh := cli.Subscribe(ctx, subscribeFilters...)
+
+	if err := c.initializeRunningContainers(ctx, cli, ch); err != nil {
+		logrus.Errorf("failed to initialize existing containers published ports: %v", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancellation: %w", ctx.Err())
```

Comment:
I don't think this is an error:

```suggestion
			return nil
```